src/data/sparse_page_source.h

```diff
@@ -3,6 +3,37 @@
  * \file page_csr_source.h
  *  External memory data source, saved with sparse_batch_page binary format.
  * \author Tianqi Chen
+ *
+ * -------------------------------------------------
+ * Random notes on implementation of external memory
+ * -------------------------------------------------
+ *
+ * As of XGBoost 1.3, the general pipeline is:
+ *
+ * dmlc text file parser --> file adapter --> sparse page source -> data pool -->
```

Comment:
I don't like adding more layers. One of the primary issues I have working with DMatrix is too many conceptual layers. Perhaps you could perform the data poolling at the file adapter level? It makes sense to me that the file adapter would adapter the input into a consistent format for xgboost.

Reply:
I thought about it, it's possible to add it to file adapter but not as straight forward.  The data pool here is the consumer of data.  Imagine it as a water pool, when the pool is full it can discharge proactively (not sure if that's the right term, hope you understand).  File adapter on the other hand, it's a provider of data, it only transfer the data when it's asked to do so.

Reply:
Perhaps I can move the pooling into file adapter as a intermediate layer between the `Value` function and parser.  Feel free to suggest.

Reply:
Yes that's roughly what I had in mind. This encapsulates it and means we don't have to think about it anywhere else in xgb.